### PR TITLE
Migrate to TypeScript

### PR DESCRIPTION
**Pull request by acrylic-style**
For better typing :)
Also, modified .eslintrc file to process .ts files.

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -1,6 +1,10 @@
...
-    "@babel/preset-env"
+    "@babel/preset-env",
+    "@babel/preset-typescript"
   ],
-  "plugins": ["@babel/plugin-transform-async-to-generator"]
+  "plugins": [
+    "@babel/plugin-transform-async-to-generator",
+    "@babel/plugin-proposal-class-properties"
+  ]
 }
```
Comment (oamaok) [inline .babelrc:10]: We can ditch `babel` altogether if using typescript.

File: package.json [hunk 2]
```diff
@@ -45,6 +49,7 @@
     "regenerator-runtime": "^0.13.7",
     "sass": "^1.32.8",
     "sass-loader": "^11.0.1",
+    "typescript": "^4.2.3",
```
Comment (acrylic-style) [inline package.json:52]: New packages were added, so you need to do `yarn` after this PR.

**Comment by oamaok** [comment]
`tsconfig.json` seems to be missing, I suggest we use a relatively strict one:
```
{
  "compilerOptions": {
    "esModuleInterop": true,
    "module": "commonjs",
    "noImplicitAny": true,
    "noImplicitThis": true,
    "removeComments": true,
    "preserveConstEnums": true,
    "strictFunctionTypes": true,
    "strictNullChecks": true,
    "sourceMap": true,
    "jsx": "react",
    "moduleResolution": "node"
  }
}
```

File: .babelrc (hunk 1 above)
Comment (acrylic-style) [reply .babelrc:10]: `yarn start:chrome` threw an error (it said requires plugin-proposal-class-properties) without it for me, so it's there

File: popup/calculators/taiko.ts [hunk 3]
```diff
@@ -123,14 +102,16 @@ export const calculate = (map, mods, taikoObjects) => {
...
+    parsedTaikoResult,
+    clockRate,
+    convert
   )
   const sectionLength = 400 * clockRate
   let currentSectionEnd =
     Math.ceil(map.objects[0].time / sectionLength) * sectionLength
 
   difficultyHitObjects.forEach((h) => {
+    console.log(h)
```
Comment (oamaok) [inline popup/calculators/taiko.ts:114]: Can you add a linter rule for banning `console.log` statements? https://eslint.org/docs/rules/no-console

Comment (acrylic-style) [reply popup/calculators/taiko.ts:114]: yeah, I forgot to remove this after debugging it

Comment (acrylic-style) [reply popup/calculators/taiko.ts:114]: https://github.com/oamaok/ezpp/blob/2c2cc29c1a7b03dadd69cfc494999e4ea7051b8c/popup/analytics.js#L5
do we need this console.log call?

File: .babelrc (hunk 1 above)
Comment (oamaok) [reply .babelrc:10]: we need to change the webpack configs as well for this to work. replace the `babel-loader` there with `ts-loader`.

File: popup/calculators/taiko.ts (hunk 3 above)
Comment (oamaok) [reply popup/calculators/taiko.ts]: it's handy for analytics debugging, so let's leave it.

File: popup/converters/taiko.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+export const convertHitObjects = (
+  objects: Array<TaikoObject>,
+  map: ojsama.beatmap
+): Array<TaikoObject> => {
+  // not executed because star rating calculation for conversion maps are disabled
+  const result: Array<TaikoObject> = []
+  objects.forEach((obj) => {
+    convertHitObject(obj, map).forEach((e) => {
+      result.push(e)
```
Comment (oamaok) [inline popup/converters/taiko.ts:15]:
```ts
const result: Array<TaikoObject> = objects.flatMap(obj => convertHitObject(obj, map))
```

File: popup/translations.ts [hunk 5]
```diff
@@ -40,15 +49,20 @@ export const getTranslation = (translationKey, ...args) => {
...
 ) => {
   if (setterHooks.some((hook) => hook.element === element)) {
     throw new Error('This element already has a text setter')
   }
 
-  const hook = {
+  const hook: {
+    element: Element
+    translationKey: string
+    property: string
```
Comment (oamaok) [inline popup/translations.ts:63]: The type here should be `property: keyof HTMLElement` or `property: keyof Element`

File: popup/translations.ts [hunk 6]
```diff
@@ -14,18 +14,27 @@ languages
     languageSelector.appendChild(option)
   })
 
-export const translations = languages.reduce(
+export const translations: {
+  [code: string]: {
+    [key: string]: string
+  }
```
Comment (oamaok) [inline popup/translations.ts:20]: Let's use TypeScript's utility types for records: `Record<string, Record<string, string>>`

File: popup/skills/taiko/staminaCheeseDetector.ts [hunk 7]
```diff
@@ -2,17 +2,17 @@ import { HitType } from '../../objects/taiko/hitType'
 import TaikoDifficultyHitObject from '../../objects/taiko/taikoDifficultyHitObject'
 import LimitedCapacityQueue from '../../util/limitedCapacityQueue'
 
+export const roll_min_repetitions = 12
+export const tl_min_repetitions = 16
```
Comment (oamaok) [inline popup/skills/taiko/staminaCheeseDetector.ts:6]: Constants should be defined in all capital case, `TL_MIN_REPETITIONS`

File: popup/skills/taiko/stamina.ts [hunk 8]
```diff
@@ -1,70 +1,60 @@
...
     this.hand = rightHand ? 1 : 0
   }
 
-  /**
-   * @param {DifficultyHitObject} current
-   */
-  strainValueOf(current) {
-    if (!(current.baseObject.type & ObjectType.Hit)) {
+  public strainValueOf(current: TaikoDifficultyHitObject): number {
+    if (!(current.baseObject.objectType === ObjectType.Hit)) {
```
Comment (oamaok) [inline popup/skills/taiko/stamina.ts:23]: `current.baseObject.objectType !== ObjectType.Hit`

File: popup/skills/taiko/rhythm.ts [hunk 9]
```diff
@@ -8,41 +7,37 @@ export const STRAIN_DECAY = 0.96
...
-    this.notesSinceRhythmChange = 0
   }
 
-  /**
-   * @param {DifficultyHitObject} current
-   */
-  strainValueOf(current) {
-    if (!(current.baseObject.type & ObjectType.Hit)) {
+  public strainValueOf(current: TaikoDifficultyHitObject): number {
+    if (!(current.baseObject.objectType === ObjectType.Hit)) {
```
Comment (oamaok) [inline popup/skills/taiko/rhythm.ts:23]: `current.baseObject.objectType !== ObjectType.Hit`

File: popup/settings.ts [hunk 10]
```diff
@@ -25,18 +25,22 @@ const SETTINGS = [
   },
 ]
 
-let currentSettings = {}
-const settingsChangeListeners = []
+let currentSettings: {
+  [key: string]: any
```
Comment (oamaok) [inline popup/settings.ts:29]: `let currentSettings: Record<string, any>`

File: popup/calculators/taiko.ts [hunk 11]
```diff
@@ -242,7 +216,7 @@ export const locallyCombinedDifficulty = (
       (staminaRight.strainPeaks[i] + staminaLeft.strainPeaks[i]) *
       STAMINA_SKILL_MULTIPLIER *
       staminaPenalty
-    peaks.push(norm(2, [colourPeak, rhythmPeak, staminaPeak]))
+    peaks.push(norm(2, colourPeak, rhythmPeak, staminaPeak))
```
Comment (oamaok) [inline popup/calculators/taiko.ts:219]:
```ts
const peaks = colour.strainPeaks.map(peak => {
  ...
  return norm(2, colourPeak, rhythmPeak, staminaPeak)
})
```

File: popup/calculators/taiko.ts [hunk 12]
```diff
@@ -16,69 +18,38 @@ export const COLOUR_SKILL_MULTIPLIER = 0.01
...
-  const rawTaikoObjects = []
+export const createDifficultyHitObjects = (
+  map: beatmap,
+  parsedTaikoResult: ParsedTaikoResult,
+  clockRate: number,
+  convert: boolean
+) => {
+  const rawTaikoObjects = [] as Array<TaikoObject>
   for (let i = 0; i < map.objects.length; i++) {
     rawTaikoObjects.push(
```
Comment (oamaok) [inline popup/calculators/taiko.ts:40]:
```ts
const rawTaikoObjects: TaikoObject[] = map.objects.map(obj => new TaikoObject( ... ))
```

**Review by oamaok — CHANGES_REQUESTED** [review]
Just some minor issues, looking really good overall! There are also many other places where we can replace `for` loops with array methods, but those can be improved later.

**Comment by acrylic-style** [comment]
Let me fix more `for` loops before merging it

File: popup/translations.ts [hunk 13]
```diff
@@ -57,13 +70,14 @@ export const createTextSetter = (
 
   setterHooks.push(hook)
 
-  return (...args) => {
+  return (...args: any[]) => {
     hook.args = args
+    // @ts-ignore
```
Comment (oamaok) [inline popup/translations.ts:75]: you should be able to remove this now that the `property` is safely a member of `element`

File: popup/translations.ts [hunk 14]
```diff
@@ -14,18 +14,26 @@ languages
...
   {}
 )
 
-let currentLanguage
-const setterHooks = []
+let currentLanguage: string
+const setterHooks: {
+  element: Element
+  translationKey: string
+  property: string
```
Comment (oamaok) [inline popup/translations.ts:32]: this should still be `keyof HTMLElement`

File: popup/translations.ts [hunk 15]
```diff
@@ -74,10 +88,11 @@ export const setLanguage = (language) => {
   currentLanguage = language
 
   setterHooks.forEach(({ element, translationKey, property, args }) => {
+    // @ts-ignore
```
Comment (oamaok) [inline popup/translations.ts:91]: this can also be removed once the `property` is properly typed

File: popup/translations.ts (hunk 13 above)
Comment (acrylic-style) [reply popup/translations.ts:75]: image

Comment (oamaok) [reply popup/translations.ts:75]: hmm, the element should be then typed as `HTMLElement` as well

Comment (acrylic-style) [reply popup/translations.ts:75]: It still complains about read-only properties
image

Comment (oamaok) [reply popup/translations.ts:75]: ahh, damn. I'm pretty sure there's a TypeScript trick that can be used here. let me take a quick look.

Comment (oamaok) [reply popup/translations.ts:75]: https://stackoverflow.com/questions/49579094/typescript-conditional-types-filter-out-readonly-properties-pick-only-requir

The solution is quite complex and maybe not worth tackling right now. Let's just keep the `@ts-ignore` there for now.

Comment (oamaok) [reply popup/translations.ts:75]: Alternate, quick solution: just type the property with this:
```
type TranslatableProperties = 'innerHTML'

property: TranslatableProperties
```

this way we can extend the attributes there easily, and it's type safe.

**Review by oamaok — APPROVED** [review]
Let's go!

**Pull request merged**